File: .toys/update-disco.rb
# frozen_string_literal: true

# Copyright 2022 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

desc "Update the discovery documents."

long_desc \
  "This tool downloads the latest discovery documents from the discovery " \
  "service, and updates the cache in the discovery directory. It debounces " \
  "revision changes to prevent switching to an earlier revision, and also " \
  "ignores revisions that are identical except for the revision number.",
  "",
  "By default, this tool just makes the changes locally. To open a pull " \
  "request with the changes, set the --remote= and/or --fork arguments. " \
  "To auto-approve and automerge, set the --approval-token= argument or " \
  "the APPROVAL_GITHUB_TOKEN environment variable."

flag :git_remote, "--remote=NAME" do
  desc "The name of the git remote to use as the pull request head."
end
flag :enable_fork, "--fork" do
  desc "The github user for whom to create/use a fork."
end
flag :approval_token, "--approval-token=TOKEN" do
  default ENV["APPROVAL_GITHUB_TOKEN"]
  desc "GitHub token for adding labels and approving pull requests. Defaults " \
       "to the value of the APPROVAL_GITHUB_TOKEN environment variable."
end

include :exec, e: true
include :fileutils
include "yoshi-pr-generator"

def run
  setup
  result = open_pr
  output result
end

def setup
  Dir.chdir context_directory
  @timestamp = Time.now.utc.strftime("%Y%m%d-%H%M%S")
  yoshi_utils.git_ensure_identity
  if enable_fork
    set :git_remote, "pull-request-fork" unless git_remote
    yoshi_utils.gh_ensure_fork remote: git_remote
  end
  require "net/http"
  require "json"
  require "set"
end

def open_pr
  branch_name = "disco-#{@timestamp}"
  commit_message = "chore: Automated update of discovery documents at #{@timestamp}"
  approval_message = "Rubber-stamped automated update of discovery documents!"
  yoshi_pr_generator.capture enabled: !git_remote.nil?,
                             remote: git_remote,
                             branch_name: branch_name,
                             commit_message: commit_message,
                             labels: ["automerge"],
                             auto_approve: approval_message,
                             approval_token: approval_token do
    update_disco
  end
end

class DocumentInfo
  def initialize content, filename = nil
    @filename = filename || "index.json"
    @content = content
    @json = JSON.parse content rescue nil
    @revision = @json && @json["revision"]
  end

  attr_reader :filename
  attr_reader :content
  attr_reader :json
  attr_reader :revision

  def json_without_revision
    @json_without_revision ||= begin
      result = @json.dup
      result.delete "revision"
      result.delete "etag"
      result
    end
  end
end

def update_disco
  Dir.chdir "discoveries" do
    index_document = load_index
    documents = load_documents index_document
    delete_unused_files index_document, *documents
    update_files documents
    update_index index_document
  end
end

def load_index
  logger.unknown "Loading index ..."
  content = Net::HTTP.get URI("https://discovery.googleapis.com/discovery/v1/apis") rescue nil
  document = DocumentInfo.new content
  unless document.json
    logger.fatal "Unable to load discovery index", :red, :bold
    exit 1
  end
  logger.info "Loaded discovery index"
  document
end

def load_documents index_document
  logger.unknown "Loading discovery documents ..."
  result = []
  index_document.json["items"].each do |item|
    name = item["name"]
    version = item["version"]
    discovery_rest_url = item["discoveryRestUrl"]
    filename = "#{name}.#{version}.json"
    content = Net::HTTP.get URI(discovery_rest_url) rescue nil
    document = DocumentInfo.new content, filename if content
    if document.revision
      logger.info "Loaded #{filename} from #{discovery_rest_url}"
      result << document
    else
      logger.error "Failed to load #{filename} from #{discovery_rest_url}"
    end
  end
  result
end

def delete_unused_files *documents
  expected_names = Set.new documents.map(&:filename)
  Dir.glob "*.json" do |filename|
    next if expected_names.include? filename
    logger.unknown "REMOVING file #{filename}"
    rm_rf filename
  end
end

def update_files documents
  documents.each do |document|
    filename = document.filename
    existing_content = File.read filename rescue nil
    existing = DocumentInfo.new existing_content, filename
    if existing.revision
      if existing.revision > document.revision
        logger.info "Existing revision #{existing.revision} > updated revision #{document.revision} for #{filename}"
        next
      elsif existing.revision == document.revision
        logger.info "No change to revision #{existing.revision} for #{filename}"
        next
      elsif existing.json_without_revision == document.json_without_revision
        logger.info "Revision updated #{existing.revision} to #{document.revision} but no other change for #{filename}"
        next
      end
      logger.unknown "UPDATING revision #{existing.revision} to #{document.revision} for #{filename}"
    else
      logger.unknown "WRITING new revision #{document.revision} for #{filename}"
    end
    File.write filename, document.content
  end
end

def update_index index_document
  existing_content = File.read index_document.filename
  if existing_content == index_document.content
    logger.info "Index is unchanged"
  else
    logger.unknown "UPDATING index.json"
    File.write index_document.filename, index_document.content
  end
end

def output result
  case result
  when Integer
    logger.unknown "Opened pull request #{result} for discovery documents at #{@timestamp}"
  when :unchanged
    logger.unknown "No changes for discovery documents at #{@timestamp}"
  else
    logger.unknown "Updated discovery documents at #{@timestamp} but did not open a pull request"
  end
end

File: scripts/update_disco.py
# Copyright 2022 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import glob
import json
import logging
import os
import os.path
import sys
import urllib.request
from typing import Any, Optional


def main() -> None:
    """The entrypoint for updating discovery documents

    This script pulls the latest discovery documents from the discovery service
    and makes any necessary changes to update the discoveries directory.
    """
    logging.basicConfig(level=logging.INFO)
    os.chdir("discoveries")
    index_document: DocumentInfo = load_index()
    service_documents: list[DocumentInfo] = load_documents(index_document)
    delete_unused_files(index_document, service_documents)
    update_files(service_documents)
    update_index(index_document)
    os.chdir("..")


class DocumentInfo:
    """An object that provides information about a discovery document

    This could be a discovery document or a discovery index. The former will
    have a revision string, whereas the latter will set it to None.

    Attributes:
        filename {str} -- The name of the json file in the discoveries dir
        content {bytes} -- The raw source
        json {object} -- The parsed JSON object, or None if parsing failed
        revision {str} -- The document revision, or None if this is an index
        json_without_revision {object} -- The parsed JSON object with revision
            and etag fields removed, for semantic comparisons
        json_string {str} -- The JSON object serialised to string, with keys sorted
    """

    filename: str
    content: bytes
    json: Any
    revision: Optional[str]
    json_without_revision: Any
    json_string: str

    def __init__(self, content: bytes, filename: Optional[str] = None) -> None:
        """Analyzes document content

        Arguments:
            content {bytes} -- The raw content from a file or HTTP response
            filename {str} -- The filename (defaults to the index filename)
        """
        self.filename = filename or "index.json"
        self.content = content
        try:
            self.json = json.loads(content)
            self.revision = self.json.get("revision")
            self.json_without_revision = self.json.copy()
            self.json_string = json.dumps(self.json, indent=2, sort_keys=True)
            if "revision" in self.json_without_revision:
                del self.json_without_revision["revision"]
            if "etag" in self.json_without_revision:
                del self.json_without_revision["etag"]
        except json.decoder.JSONDecodeError:
            self.json = None
            self.json_without_revision = None
            self.revision = None
            self.json_string = ""


def load_index() -> DocumentInfo:
    """Downloads the current discovery index and returns a document

    Returns:
        DocumentInfo -- Analysis of the discovery index content
    """
    print("LOADING index document ...")
    with urllib.request.urlopen(
        "https://discovery.googleapis.com/discovery/v1/apis"
    ) as f:
        if f.status != 200:
            sys.exit(f"Got HTTP status {f.status} for discovery index")
        document: DocumentInfo = DocumentInfo(f.read())
    if document.json is None:
        sys.exit("Unable to parse discovery index")
    logging.info("Loaded index")
    return document


def load_documents(index_document: DocumentInfo) -> list[DocumentInfo]:
    """Downloads all discovery documents in the index

    Arguments:
        index_document {DocumentInfo} -- The index

    Returns:
        list[DocumentInfo] -- List of discovery document content
    """
    print("LOADING service documents ...")
    service_documents: list[DocumentInfo] = []
    for item in index_document.json["items"]:
        name: str = item["name"]
        version: str = item["version"]
        discovery_rest_url: str = item["discoveryRestUrl"]
        filename: str = f"{name}.{version}.json"
        # Sometimes the index lists services that don't exist. So log any
        # errors but don't let them crash the entire script.
        try:
            with urllib.request.urlopen(discovery_rest_url) as f:
                if f.status != 200:
                    logging.error(
                        f"HTTP status {f.status} when loading {filename} from {discovery_rest_url}"
                    )
                    continue
                document: DocumentInfo = DocumentInfo(f.read(), filename)
        except urllib.error.HTTPError:
            logging.error(f"Failed to load {filename} from {discovery_rest_url}")
            continue
        if document.revision == None:
            logging.error(
                f"Malformed document for {filename} from {discovery_rest_url}"
            )
            continue
        logging.info(f"Loaded {filename} from {discovery_rest_url}")
        service_documents.append(document)
    return service_documents


def delete_unused_files(
    index_document: DocumentInfo, service_documents: list[DocumentInfo]
) -> None:
    """Deletes files in the discoveries directory that are no longer used

    Arguments:
        index_document {DocumentInfo} -- The index
        service_documents {Sequence[DocumentInfo]} -- The discoveries
    """
    expected_names: set[str] = set([index_document.filename])
    for doc in service_documents:
        expected_names.add(doc.filename)
    for filename in glob.glob("*.json"):
        if filename in expected_names:
            continue
        print(f"REMOVING file {filename}")
        os.remove(filename)


def update_files(service_documents: list[DocumentInfo]) -> None:
    """Iterates the service discovery documents and updates the discoveries dir

    This updates only files where we downloaded a revision strictly greater
    than the existing revision, AND whose content has semantically changed.

    Arguments:
        service_documents {list[DocumentInfo]} -- The discoveries
    """
    for document in service_documents:
        document_revision: Optional[str] = document.revision
        assert document_revision is not None
        filename: str = document.filename
        if os.path.isfile(filename):
            with open(filename, "rb+") as f:
                existing: DocumentInfo = DocumentInfo(f.read(), filename)
                existing_revision: str = existing.revision or "(unknown)"
                if existing_revision > document_revision:
                    logging.info(
                        f"Existing revision {existing_revision} > updated revision {document_revision} for {filename}"
                    )
                    continue
                elif existing_revision == document_revision:
                    logging.info(
                        f"No change to revision {document_revision} for {filename}"
                    )
                    continue
                elif existing.json_without_revision == document.json_without_revision:
                    logging.info(
                        f"Revision updated {existing_revision} to {document_revision} but no other changes for {filename}"
                    )
                    continue
                print(
                    f"UPDATING revision {existing_revision} to {document_revision} for {filename}"
                )
                f.seek(0)
                f.truncate()
                f.write(document.json_string.encode())
        else:
            print(f"WRITING new file {filename} at revision {document_revision}")
            with open(filename, "wb") as f:
                f.write(document.json_string.encode())


def update_index(index_document: DocumentInfo) -> None:
    """Updates the index file in the discoveries dir

    Arguments:
        index_document {DocumentInfo} -- The index
    """
    filename = index_document.filename
    with open(filename, "rb+") as f:
        existing_content = f.read()
        if existing_content == index_document.content:
            logging.info(f"Index file {filename} is unchanged")
        else:
            print(f"UPDATING index file {filename}")
            f.seek(0)
            f.truncate()
            f.write(index_document.json_string.encode())


if __name__ == "__main__":
    main()
